lib/web_ui/dev/test_runner.dart

```diff
@@ -448,69 +480,69 @@ class TestCommand extends Command<bool> with ArgUtils {
     timestampFile.writeAsStringSync(timestamp);
   }
 
+  Future<void> _buildTestsInParallel(
+      {List<FilePath> targets, bool forCanvasKit = false}) async {
+    final List<TestBuildInput> buildInputs = targets
+        .map((FilePath f) => TestBuildInput(f, forCanvasKit: forCanvasKit))
+        .toList();
+
+    final results = _pool.forEach(
+      buildInputs,
+      _buildTest,
+    );
+    await for (final bool isSuccess in results) {
+      if (!isSuccess) {
+        throw ToolException('Failed to compile tests.');
+      }
+    }
+  }
+
   /// Builds the specific test [targets].
   ///
   /// [targets] must not be null.
   ///
-  /// When building for CanvasKit we have to use a separate `build.canvaskit.yaml`
-  /// config file. Otherwise, `build.html.yaml` is used. Because `build_runner`
-  /// overwrites the output directories, we redirect the CanvasKit output to a
-  /// separate directory, then copy the files back to `build/test`.
-  Future<void> _buildTests({List<FilePath> targets, bool forCanvasKit}) async {
-    print(
-        'Building ${targets.length} targets for ${forCanvasKit ? 'CanvasKit' : 'HTML'}');
-    final String canvasKitOutputRelativePath =
-        path.join('.dart_tool', 'canvaskit_tests');
+  /// Uses `dart2js` for building the tests.
+  ///
+  /// When building for CanvasKit we have to use extra argument
+  /// `DFLUTTER_WEB_USE_SKIA=true`.
+  Future<bool> _buildTest(TestBuildInput input) async {
+    final targetFileName = input.target.relativeToWebUi
+        .replaceFirst('.dart', '.dart.browser_test.dart.js');
+    final String targetPath = path.join('build', targetFileName);
+
+    final io.Directory directoryToTarget = io.Directory(path.join(
+        environment.webUiBuildDir.path,
+        path.dirname(input.target.relativeToWebUi)));
+
+    if (!directoryToTarget.existsSync()) {
+      directoryToTarget.createSync(recursive: true);
+    }
+
     List<String> arguments = <String>[
-      'run',
-      'build_runner',
-      'build',
+      '--no-minify',
+      '--disable-inlining',
+      '--enable-asserts',
       '--enable-experiment=non-nullable',
-      'test',
+      '--no-sound-null-safety',
+      if (input.forCanvasKit) '-DFLUTTER_WEB_USE_SKIA=true',
+      '-O2',
       '-o',
-      forCanvasKit ? canvasKitOutputRelativePath : 'build',
-      '--config',
-      // CanvasKit uses `build.canvaskit.yaml`, which HTML Uses `build.html.yaml`.
-      forCanvasKit ? 'canvaskit' : 'html',
-      for (FilePath path in targets) ...[
-        '--build-filter=${path.relativeToWebUi}.js',
-        '--build-filter=${path.relativeToWebUi}.browser_test.dart.js',
-      ],
+      '${targetPath}', // target path.
+      '${input.target.relativeToWebUi}', // current path.
     ];
 
     final int exitCode = await runProcess(
```

Comment:
Have you tried running a test with an intentional error in the code. Let's make sure that when the compiler fails it prints a nice error to the console.

Reply:
Yes I did. It looks like this. It is as nice as it was before 👍 

```
00:14 +475 ~2 -19: test/engine/pointer_binding_test.dart: TouchAdapter does calculate delta and pointer identifier correctly [E]                                                                                                                                                                                                                                           
  Expected: <20>
    Actual: <40>
  
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/js_helper.dart 1129:37                             Object.wrapException
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/js_helper.dart 1160:25                             Object.throwExpression
  /.pub-cache/hosted/pub.dartlang.org/test_api-0.2.15/lib/src/frontend/expect.dart 154:30             Object.fail
  /.pub-cache/hosted/pub.dartlang.org/test_api-0.2.15/lib/src/frontend/expect.dart 148:3              Object._expect
  /.pub-cache/hosted/pub.dartlang.org/test_api-0.2.15/lib/src/frontend/expect.dart 60:3               Object.expect
  ../../../test/engine/pointer_binding_test.dart 1966:7                                               testMain.<fn>
  ../../../test/engine/pointer_binding_test.dart 31:9                                                 _testEach.<fn>
  /.pub-cache/hosted/pub.dartlang.org/test_api-0.2.15/lib/src/backend/declarer.dart 172:23            <fn>
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/async_patch.dart 315:19                            _wrapJsFunctionForAsync.closure.$protected
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/async_patch.dart 340:23                            _wrapJsFunctionForAsync.<fn>
  ===== asynchronous gap ===========================
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/js_helper.dart 2182:5                              Object.eval
  org-dartlang-sdk:///lib/async/zone.dart 1128:34                                                     _CustomZone.registerBinaryCallback
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/async_patch.dart 339:23                            Object._wrapJsFunctionForAsync
  org-dartlang-sdk:///lib/async/zone.dart 1190:12                                                     StaticClosure._rootRun
  org-dartlang-sdk:///lib/async/zone.dart 1092:34                                                     _CustomZone.run
  ===== asynchronous gap ===========================
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/js_helper.dart 2182:5                              Object.eval
  org-dartlang-sdk:///lib/async/zone.dart 1128:34                                                     _CustomZone.registerBinaryCallback
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/async_patch.dart 339:23                            Object._wrapJsFunctionForAsync
  /.pub-cache/hosted/pub.dartlang.org/test_api-0.2.15/lib/src/backend/invoker.dart 400:30             <fn>
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/async_patch.dart 315:19                            _wrapJsFunctionForAsync.closure.$protected
  ===== asynchronous gap ===========================
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/js_helper.dart 2182:5                              Object.eval
  org-dartlang-sdk:///lib/async/zone.dart 1128:34                                                     _CustomZone.registerBinaryCallback
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/async_patch.dart 339:23                            Object._wrapJsFunctionForAsync
  org-dartlang-sdk:///lib/async/future.dart 175:26                                                    Future.Future.<fn>
  /.pub-cache/hosted/pub.dartlang.org/stack_trace-1.9.5/lib/src/stack_zone_specification.dart 206:14  StackZoneSpecification._run
  ===== asynchronous gap ===========================
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/js_helper.dart 2182:5                              Object.eval
  org-dartlang-sdk:///lib/async/zone.dart 1113:34                                                     _CustomZone.registerCallback
  org-dartlang-sdk:///lib/async/zone.dart 1036:22                                                     _CustomZone.bindCallbackGuarded
  org-dartlang-sdk:///lib/async/timer.dart 52:45                                                      Object.Timer.Timer
  org-dartlang-sdk:///lib/async/timer.dart 89:9                                                       Object.Timer.run
  org-dartlang-sdk:///lib/async/future.dart 173:11                                                    Object.Future.Future
  ===== asynchronous gap ===========================
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/js_helper.dart 2182:5                              Object.eval
  org-dartlang-sdk:///lib/async/zone.dart 1128:34                                                     _CustomZone.registerBinaryCallback
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/async_patch.dart 339:23                            Object._wrapJsFunctionForAsync
  org-dartlang-sdk:///lib/async/zone.dart 1190:12                                                     StaticClosure._rootRun
  org-dartlang-sdk:///lib/async/zone.dart 1092:34   
```

Reply:
Ah I see you are asking for a compile error :D Here it is:

```
test/alarm_clock_test.dart:19:5:
Error: Method not found: '_alarmCloTests'.
    _alarmCloTests();
    ^^^^^^^^^^^^^^
Error: Compilation failed.
ERROR: Failed to compile test /Users/nurhan/development/engine/src/flutter/lib/web_ui/test/alarm_clock_test.dart. Dart2js exited with exit code 1
Failed to compile tests.
```

Reply:
Awesome! Thanks for checking.